### ISSUE TITLE
An inferred signature anchors on a negative-test call site and models a tuple parameter as a single `TensorSpec`

### ISSUE THREAD

**Issue by khatchad** [ponder-lab/Hybridize-Functions-Refactoring#888]
## What Is Emitted

At 0.52.83, the signature inferred for `CIN.call`'s `inputs` parameter in the deep_recommenders subject (`deep_recommenders/keras/models/ranking/xdeepfm.py:71`) is a single `TensorSpec` of `float32 (2, 3, 5)`. The parameter is not a tensor: every conforming call site passes a two-tensor tuple `(x0, x)`, unpacked at `xdeepfm.py:73`.

## Where The Shape Comes From

`(2, 3, 5)` is exactly the argument of `tests/keras/test_xdeepfm.py:21`, a negative test: `test_invalid_inputs_type` passes a bare `(2, 3, 5)` ndarray inside `assertRaisesRegexp` (:18-20) so the layer's own type guard raises (`xdeepfm.py:41-43`). The inference anchored the signature on a call site whose purpose is to be rejected by the function it types. The second negative test at :28 (a rank-2 tuple asserting the ndim guard at `xdeepfm.py:75-77`) is the same hazard one row over.

## What The Conforming Sites Pass

Tuples of two rank-3 `float32` tensors throughout: `((1, 2, 3), (1, 2, 3))` eager pairs (`test_xdeepfm.py:32-37`, `:47-54`), `Input(shape=(12, 10))` pairs executed with `(10, 12, 10)` batches (`:66-68` with `:74`/`:79`, `:85-87` with `:93`/`:96`), and the chained second `CIN` receiving the previous layer's `(10, 3, 10)` output. Every axis of both tuple elements varies across these sites.

## Two Defects, Possibly Separable

1. **Negative-test anchoring.** A call site inside an expected-failure assertion contributes to the signature. Any signature derived from it can only admit the input the body itself rejects; a decorated `CIN` would refuse every conforming caller while accepting exactly the argument that raises. Whether the fix is recognizing `assertRaises` contexts or discounting call sites that reach the function's own raise path, the emitted spec should not be derivable from a call the function is specified to reject.
2. **A tuple parameter emitted as one `TensorSpec`.** Even with the anchor corrected, the parameter needs a nested structure (two specs), not a single spec. The same mismodeling appears without the negative-test confound in the TensorFlow2.0-Examples subject: `train_step`'s and `distributed_train_step`'s `(images, labels)` tuple parameters in `7-Utils/multi_gpu_train.py` each emit a single unknown-rank `float32` spec. Related to the closed #790, which was the inverse classification defect.

**Comment by khatchad** [comment]
State after #897, so this issue stays self-contained.

## What #897 Does

The emission ceiling described in the 2026-08-09 comment is gone. Where a parameter's every observed tensor type comes from a declared expected failure, the container question is now asked of the conforming nodes alone, and the nested spec those callers support is emitted instead of nothing. The reduced anchor recovers: `cin` in `testInputSignatureMixedContainerEvidence` emits the same two-element spec as its `tuples_only` control, which reaches that spec with nothing excluded anywhere.

Both halves of the container evidence had to leave the guarded node out, the boolean verdict and the element extraction alike, since the bare tensor that node passes is neither a sequence nor element evidence and beside the tuple would report the form unsupported.

The #890 lesson was carried in as intended: whole-project verification ran before the merge rather than after, and against a baseline built from `main` rather than against the accepted output alone, so the result is attributable to the change and not to what has landed since. It moves nothing at that scale.

## Why This Issue Stays Open

`CIN.call` does not recover. Its container check answers no: no tensor container is associated with `inputs` at all in the subject, so there is no element structure for the reduction to fall through to, and the parameter still withholds with `EXPECTED_FAILURE_EVIDENCE_ONLY`. The reduced fixture recovers precisely because its tuple *is* recognized as one.

So what is left here is the evidence rather than the dispatch, which is the same shape as the `train_step` half already sent upstream as wala/ML#830. The remaining work is to establish why the tuple reaching a Keras layer's `call` is not associated with the parameter as a container, and to take that upstream if it belongs there.

## Related, Not Part Of This

Building the arm that keeps the withholding pinned turned up #898: a guard is read as covering every later call in its frame rather than its own `with` body. That is a defect in the exclusion #892 introduced here, but it is independent of what remains open above, and neither one blocks the other.

**Comment by khatchad** [comment]
State, and what is actually left.

## The First Defect Is Settled

Negative-test anchoring is resolved, and not in the way the report expected. #892 stopped a specification being derived from a call the tests declare must fail, and #897 recovers the nested spec the conforming callers support where the evidence exists. #904 then removed the last thing standing between an annotated parameter and that recovery.

The reported `CIN.call` ends up withholding regardless, and correctly: it declares `**kwargs`, so since #903 no signature is written for it at all. An input signature fixes the arguments a function accepts, the rest-keyword slot then absorbs nothing, and Keras passing `training` breaks the layer. The emission this issue asked for is one the tool is now right to refuse, which is a better answer than the one requested.

## The Second Defect Is Fixed Upstream And Not Yet Reachable

The `train_step` half, the `(images, labels)` batch tuple in TensorFlow2.0-Examples' `7-Utils/multi_gpu_train.py` that emits a single unknown-rank `float32` spec, was diagnosed as a modeling matter rather than a dispatch one and sent upstream. That work is done there.

It is not reachable from here yet: this repository pins `com.ibm.wala.cast.python.ml` at 0.52.83, and the fix is in no released version. Nothing in this repository can move that.

## The Next Step

Not a code change here. In order:

1. An Ariadne release carrying the upstream fix.
2. A version bump in this repository, under the usual release-verification loop.
3. Verification that `train_step` and `distributed_train_step` emit the two-element nested spec rather than the single unknown-rank one. That is the evidence this issue closes on, and it is a direct check rather than an inference, since the shape is named in the report.

Worth noting for whoever runs step 3 that the same release carries a second change to emissions, the union for parameters whose consumers impose disagreeing dtypes, so the bump's verification will be reading two moving parts at once and should attribute them separately.

**Comment by khatchad** [comment]
Reopening: the tuple half of this issue is not solved. Measured on 0.52.85, which was expected to carry it.

## The Persisting Case

A training-step function takes a batch tuple and unpacks it (`images, labels = inputs`). Its parameter reads identically on 0.52.84 and 0.52.85, byte for byte:

```
inputs   is tensor=true   is container=(blank)   container element types=(blank)
         FLOAT32[Dynamic; Constant,112; Constant,112; Constant,3] | FLOAT32[Dynamic; Unresolved]
```

Those two union members are the tuple's two elements, collapsed onto one parameter rather than separated into positions. The blank in the container column is the tell: it means the container question was never *asked*, as distinct from asked and answered no.

The emission that follows is the one this issue is about. Before, the function was given a single unknown-rank specification, which raises `OperatorNotAllowedInGraphError` because an unknown-rank tensor cannot be unpacked. It is now withheld instead, so the function converts with a bare decorator and runs. The break is gone; the nested specification this issue asks for is still not written.

## Where The Gap Is, Measured

The evidence exists and is catalogued. Probing the parameter directly at the real witness:

```
pts type=<PythonLoader,Ltuple>  container=true  inCatalog=true
fields={0=[{[Dynamic, 112, 112, 3] of float32}],
        1=[{[Dynamic, Unresolved] of float32}]}
key=SITE_IN_NODE{synthetic <...flow_from_directory...>:Ltuple:NEW <PythonLoader,Ltuple>@5 ...}
```

The tuple allocation is in the parameter's points-to set, and its instance-field keys carry both element types, correctly split by position. So the underlying analysis delivers exactly what the reduction needs.

This consumer never asks for it. Classification returns on the parameter's own tensor types:

```java
// Phase 2
if (!this.getTensorTypes().isEmpty()) {
    if (this.getConformingTensorTypes().isEmpty()) {
        this.tensorContainer = this.hasTensorContainer(...);   // asked only here
        ...
    }
    return this.tensor = TRUE;                                  // returns
}
// Phase 3: check for containers of tensors                     // never reached
```

The container question is put only when every observed type came from a call site declared to fail, which is the case #892 added. Here the evidence conforms, so the branch is skipped and Phase 3 is unreachable. The union reported on the parameter *is* the element types flattened, which is why the parameter looks well-typed and nothing appears wrong.

This is the same shape as the type-hint short-circuit #904 removed from Phase 1, still present in Phase 2's main path.

## What The Fix Is Not

Removing the short-circuit outright would put the container question to every already-tensor-typed parameter, which is a change with reach rather than a local repair. It also touches the classification path that #909's residual decline and #875 are both queued to edit, so the three want sequencing rather than three independent passes over the same predicate.

Two questions to settle before implementing, in this order:

1. What narrows the question so it is asked where it can pay. Asking whenever the parameter's instance keys include a catalogued container is the obvious candidate, since that is exactly the condition under which an answer exists.
2. Whether this rides with the #875 and #909(b) unit, which edits the same predicate from the other side, or lands first as its own change.

Nothing is owed upstream. The analysis side delivers the evidence at the real witness, and the reduction-level mechanism was independently verified to survive the indirect dispatch this witness uses.
